adoc/chapters/programming_interface.adoc

```diff
@@ -19401,481 +19401,1161 @@ that all of the built-in functions fulfill the same requirements for both
-[code]#marray# specialization.
+The functions defined in sections <<sec:math-functions>> through
+<<sec::relational-functions>> are specified using a set of generic type names.
+These are not actual type names available to SYCL applications.  Instead, they
+are a shorthand used in this document to describe the set of overloads for
+these functions.  For example the [code]#fmax# function is specified as:
 
+[source]
+----
+genfloat fmax(genfloat x, genfloat y)
+----
+
+The generic type [code]#genfloat# is a stand-in for a set of a actual type
+names: [code]#float#, [code]#double#, etc.  To construct the actual overloads
+that are available, replace the generic type name with the name of each of
+these actual types:
+
+[source]
+----
+float fmax(float x, float y)
+double fmax(double x, double y)
+vec<float,1> fmax(vec<float,1> x, vec<float,1> y)
+vec<float,2> fmax(vec<float,2> x, vec<float,2> y)
+vec<float,3> fmax(vec<float,3> x, vec<float,3> y)
+vec<float,4> fmax(vec<float,4> x, vec<float,4> y)
+vec<float,8> fmax(vec<float,8> x, vec<float,8> y)
+vec<float,16> fmax(vec<float,16> x, vec<float,16> y)
+template<size_t N> marray<float,N> fmax(marray<float,N> x, marray<float,N> y)
+// etc.
+----
+
+When determining the set of overloads that are available, use the following
+rules:
+
+* If the specification of a function has several occurrences of the same
+  generic type name, replace each occurrence with the same actual type.  For
+  example, the [code]#fmax# example above has three occurrences of
+  [code]#genfloat#: one for the return type and one for each parameter type.
+  The available overloads include only the cases where both parameters have
+  the same type, and the return type always matches the type of the parameters.
+
+* The following generic type names sometimes appear together in the same
+  function description:
++
+--
+** [code]#vfloatn#
+** [code]#vdoublen#
+** [code]#vhalfn#
+** [code]#vint8n#
+** [code]#vint16n#
+** [code]#vint32n#
+** [code]#vint64n#
+** [code]#vuint8n#
+** [code]#vuint32n#
+** [code]#vuint64n#
+** [code]#vfloatnptr#
+** [code]#vdoublenptr#
+** [code]#vhalfnptr#
+** [code]#vint32nptr#
+--
++
+When this occurs, the replacement of the generic type names are correlated.
+If one of these generic type names is replaced with a [code]#vec# of length
+_Len_, then all other occurrences of these generic type names in that same
+function description must also be replaced with a [code]#vec# of length _Len_.
+For example, if a function takes one parameter of type [code]#vfloatn# and
+another of type [code]#vfloatnptr#, there are overloads like:
++
+[source]
+----
+func(vec<float,1>, multi_ptr<vec<float,1>, Space, IsDecorated>)
+func(vec<float,2>, multi_ptr<vec<float,2>, Space, IsDecorated>)
+// ...
+----
++
+However, there are no overloads where the two parameters have different
+vector lengths.
+
+* Some of the generic type names can be replaced with a templated type, where
```

Comment:
```suggestion
* Some of the generic type names are replaced with a templated type, where
```
Should we be stronger here and say they **are** replaced with a template?

Reply:
Added in bd1229a146abf422525dd25563865087e0d2fa1a